table/multiget_context.h

```diff
@@ -0,0 +1,238 @@
+class GetContext;
+
+struct KeyContext {
+  const Slice* key;
+  ColumnFamilyHandle* column_family;
+  LookupKey* lkey;
+  Slice ukey;
+  Slice ikey;
+  Status* s;
+  MergeContext merge_context;
+  SequenceNumber max_covering_tombstone_seq;
+  bool key_exists;
+  SequenceNumber seq;
+  void* cb_arg;
+  PinnableSlice* value;
+  GetContext* get_context;
+
+  KeyContext(const Slice& user_key, ColumnFamilyHandle* col_family,
+             PinnableSlice* val, Status* stat)
+      : key(&user_key),
+        column_family(col_family),
+        lkey(nullptr),
+        s(stat),
+        max_covering_tombstone_seq(0),
+        key_exists(false),
+        seq(0),
+        cb_arg(nullptr),
+        value(val),
+        get_context(nullptr) {}
+
+  KeyContext() = default;
+};
+
+// The MultiGetContext class is a container for the sorted list of keys that
+// we need to lookup in a batch. Its main purpose is to make batch execution
+// easier by allowing various stages of the MultiGet lookups to operate on
+// subsets of keys, potentially non-contiguous. In order to accomplish this,
+// it defines the following classes -
+//
+// MultiGetContext::Range - Specifies a range of keys, by start and end index,
+// from the parent MultiGetContext. Each range contains a bit vector that
+// indicates whether the corresponding keys need to be processed or skipped.
+// A Range object can be copy constructed, and the new object inherits the
+// original Range's bit vector. This is useful for progressively skipping
+// keys as the lookup goes through various stages. For example, when looking
+// up keys in the same SST file, a Range is created excluding keys not
+// belonging to that file. A new Range is then copy constructed and individual
+// keys are skipped based on bloom filter lookup.
+//
+// MultiGetContext::Range::Iterator - A forward iterator that iterates over
+// non-skippable keys in a Range, as well as keys whose final value has been
+// found. The latter is tracked by MultiGetContext::value_mask_
+// MultiGetContext::Range::IteratorWrapper - A wrapper around a vector
+// container, such as std::vector, std::array or autovector, that shadows a
+//
+// MultiGetContext::Range::Iterator. The size of the vector must be atleast
+// the number of keys in the MultiGetContext batch (limited by
+// MultiGetContext::MAX_KEYS_ON_STACK), with the indexes of keys being
+// identical to that in MultiGetContext. This is useful for maintaining
+// auxillary data structures that can be quickly accessed while iterating
+// through a MultigetContext::Range. The auxillary data structures can be
+// allocated on stack for efficiency
+class MultiGetContext {
+ public:
+  // Limit the number of keys in a batch to this number. Benchmarks show that
+  // there is negligible benefit for batches exceeding this. Keeping this < 64
+  // simplifies iteration, as well as reduces the amount of stack allocations
+  // htat need to be performed
+  static const int MAX_KEYS_ON_STACK = 32;
```

Comment:
The total size is something like 4KB. Checked with Yoshi and he confirmed this is not a concern for MyRocks. Since the default stack allocated for each thread is ~1MB, the stack usage for MultiGet doesn't seem that large.